Fix negative pod startup duration values

ali-a-a opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
I faced an issue with the pod startup duration metric, which sometimes shows negative values. This problem is also mentioned here #130432. 

When parallel image pulling is enabled (`serializeImagePulls: false`) and the image pull policy is `IfNotPresent`, this problem may happen. When there are multiple pods concurrently pulling the image, it is possible that some of the pull requests fail (e.g., pull QPS excess). It means that some pods start recording the image pull duration, but they never stop it since they return due to an error.

https://github.com/kubernetes/kubernetes/blob/44c230bf5c321056e8bc89300b37c497f464f113/pkg/kubelet/images/image_manager.go#L284-L302

After the back-off duration, these pods retry and call `EnsureImageExists`, but this time, they don't call `pullImage` as the image already exists.

https://github.com/kubernetes/kubernetes/blob/44c230bf5c321056e8bc89300b37c497f464f113/pkg/kubelet/images/image_manager.go#L209-L246

Therefore, `state.lastFinishedPulling` never will be set.

To fix the issue, we can finish the recording by calling `RecordImageFinishedPulling` before returning from `EnsureImageExists` if there is some recording in progress (i.e., `!state.firstStartedPulling.IsZero()`). The measured pull duration is also correct, as it is the time the pod has taken to finally retrieve the image (start recording -> pull the image -> error -> retry -> image already exists -> finish recording).

I also tested the fix and couldn't reproduce the problem after that.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #130432

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ali-a-a committed:
Fix negative pod startup duration

Fixed an issue with the pod startup duration metric,
which may show negative values when parallel image pulling is enabled

bart0sh commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -210,6 +210,11 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, objRef *v1.ObjectR
+
+			// we need to stop recording if it is still in progress, as the image
+			// has already been pulled by another pod.
+			m.podPullingTimeRecorder.RecordImageFinishedPulling(pod.UID)
```
This would unconditionally update `lastFinishedPulling`, which doesn't look correct to me as in this particular case the was no pulling done (image already pulled by another pod?).

ali-a-a replied:
If `firstStartedPulling` is not set (i.e., `state.firstStartedPulling.IsZero()`), this call is a no-op. So, we wouldn't unconditionally update `lastFinishedPulling`?

If you mean we should not record the pulling time for this pod as it didn't pull the image itself, I think we should consider the time the pod tried to pull the image and failed. The `podStartSLOduration` and `podStartE2EDuration` shouldn't be the same for this pod. With my implementation, we capture the duration from the time the pod starts pulling to the time it realizes the image exists, which seems correct to me :?

bart0sh replied:
I'm still not fully understand the logic, sorry. So, some other pod downloaded the image, but for some reason (error?) it didn't update `lastFinishedPulling`. This (another) pod, doesn't download the image as image is already downloaded, but for some reason updates `lastFinishedPulling`. Why can't it be done by the pod, that does the actual downloading?
Another thing is that this image seems to be fully downloaded (without errors), but `lastFinishedPulling` is not updated. How can it happen? I'd understand this if there was a downloading error, but in this case image should be downloaded again, right?

ali-a-a replied:
> Why can't it be done by the pod, that does the actual downloading?

I think the part missing here is that each pod has its own `firstStartedPulling` and `lastFinishedPulling`. The pod that has successfully downloaded the image will update its own state:

https://github.com/kubernetes/kubernetes/blob/44c230bf5c321056e8bc89300b37c497f464f113/pkg/kubelet/util/pod_startup_latency_tracker.go#L144-149

On the other hand, pods that get an error and go into a back-off phase after setting their `firstStartedPulling`, cannot come back and update their `lastFinishedPulling`, as the image is already downloaded and the `EnsureImageExists` won't call `pullImage` again.

> Another thing is that this image seems to be fully downloaded (without errors), but `lastFinishedPulling` is not updated. How can it happen? I'd understand this if there was a downloading error, but in this case image should be downloaded again, right?

When `serializeImagePulls: false` is set, multiple pods can pull the image at the same time. When one pod successfully does that, other pods may be still in the process of downloading the image (other goroutines here are still trying to pull):

https://github.com/kubernetes/kubernetes/blob/44c230bf5c321056e8bc89300b37c497f464f113/pkg/kubelet/images/puller.go#L55-76

Again, as `lastFinishedPulling` is per pod, it is possible to fully download the image with one pod and have another pod's `lastFinishedPulling` unset.

Regarding the error itself, I also mentioned it in the PR's description. A `pull qps exceeded` error should be the reason we get those negative values. Say 20 pods are pulling an image at the same time. If the pull qps threshold is 10, 10 pods pull the image and 10 pods face the qps exceeded error. So, the error I'm referring to can be quite often if parallel pulling is enabled.

bart0sh replied:
> On the other hand, pods that get an error and go into a back-off phase after setting their firstStartedPulling, cannot come back and update their lastFinishedPulling, as the image is already downloaded and the EnsureImageExists won't call pullImage again.

Thanks for the explanations!
Would it make sense for errored pods to set `lastFinishedPulling` when error occurs?

ali-a-a replied:
It is how I think about it:

For errored pods, we could unset their `firstStartedPulling` and pretend they experienced no image pull latency (i.e., their `podStartSLOduration` == their `podStartE2EDuration`). But in this way, it doesn't match the definition we have here:

https://github.com/kubernetes/community/blob/master/sig-scalability/slos/pod_startup_latency.md

"We are explicitly excluding image pulling from time the SLI."

Imagine the errored pod gets an error in the middle of the pulling. It experiences a portion of the image-pulling duration and the backoff phase, and we are including these latencies in its SLO duration, which doesn't follow the definition.

In this implementation, in all scenarios, we measure the cost of pulling an image. For errored pods, this cost is the duration that starts with pulling and ends with realizing that the image exists on the machine after a retry. I think the final calculated duration for these pods matches the definition.

We could also define a new metric to indicate this kind of situation, but it seems like overkill.

But if you have any other ideas, I would love to try them out.

bart0sh replied:
> In this implementation, in all scenarios, we measure the cost of pulling an image. For errored pods, this cost is the duration that starts with pulling and ends with realizing that the image exists on the machine after a retry. I think the final calculated duration for these pods matches the definition.

This makes sense to me. However, there is still certain latency between the time image is successfully pulled and the time of realising this. This latency can be eliminated if `lastFinishedPulling` would be updated for all pods when the image pulling succeeds.

ali-a-a replied:
Hmm. I totally understand your point, but I have one thing on my mind. Currently, I think we consider the latency you are referring to. When one pod gets an error and retries to pull the image again, we don't reset its `firstStartedPulling`.

https://github.com/kubernetes/kubernetes/blob/44c230bf5c321056e8bc89300b37c497f464f113/pkg/kubelet/util/pod_startup_latency_tracker.go#L135-L137

 It means that the time it takes to do other things than pulling (e.g., backing off) is also included in its pull latency. Do you mean this should not be like this, and only the duration of pulling itself should be considered?

bart0sh replied:
That's a bit different story, but I understand what you mean. Let's keep your approach.

ali-a-a commented:
> @ali-a-a Thank you for your PR! Please provide unit or e2e tests that cover this issue. It's required for bugfixes.

I added a unit test for this issue, which would fail without the fix.

bart0sh commented on pkg/kubelet/container/testing/fake_runtime.go:
```diff
@@ -348,7 +347,7 @@ func (f *FakeRuntime) PullImage(ctx context.Context, image kubecontainer.ImageSp
 	}
 
-	return image.Image, retCreds, retErr
+	return image.Image, retCreds, f.Err
```
Can you explain why locking is not needed to access f.Err?

ali-a-a replied:
I'm not completely familiar with the way `fake_runtime` is used. If all the test cases run concurrently, I suppose this change could add issues. I added a new variable for the purpose of introducing errors after blocking image pull calls.

bart0sh commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -854,6 +863,88 @@ func TestMaxParallelImagePullsLimit(t *testing.T) {
+
+	// Introduce a pull error for the second pod before unblocking it
+	fakeRuntime.Err = errors.New("pull image error")
+	fakeRuntime.UnblockImagePulls(1)
```
This doesn't reproduce the error, does it? In this case image should be fully downloaded by the puller that's not errored, right? And finishedPullingRecorded should be updated by it. 
What I'd like to see is the original error, i.e. negative `podStartSLOduration`.

ali-a-a replied:
It is the reason we get negative `podStartSLOduration`. In the middle of the test, we check whether both `firstStartedPulling` values are true. At the end of the test, we check whether both `lastFinishedPulling` values are true. Without the fix, one of the `lastFinishedPulling` values wouldn't be true and essentially would cause a negative value if we wanted to calculate `podStartSLOduration` (`lastFinishedPulling (not set) - firstStartedPulling (set)`).

Could you also please look at my reply here , as I talked about it further?

https://github.com/kubernetes/kubernetes/pull/131324#discussion_r2047524590

bart0sh commented:
/ok-to-test

bart0sh commented:
/triage accepted

bart0sh commented:
@ali-a-a please fix CI failures, thanks!

bart0sh commented:
@ali-a-a Looking at `basicPodStartupLatencyTracker:ObservedPodOnWatch` code I'd say that the current approach doesn't guarantee that `podStartSLOduration` will alway be positive. For example, if it's called while image is not yet downloaded, `imagePullingDuration` and `podStartSLOduration` will be negative.

This makes me think that there should be better approach to fix https://github.com/kubernetes/kubernetes/issues/130432.

WDYT?

ali-a-a commented:
> For example, if it's called while image is not yet downloaded, `imagePullingDuration` and `podStartSLOduration` will be negative.

I cannot think of a scenario where this creates an issue and couldn't reproduce it. If `ObservedPodOnWatch` is called while the pod doesn't have the image, it won't observe the pod startup latency as the pod cannot have `hasPodStartedSLO`.

https://github.com/kubernetes/kubernetes/blob/680ea07dbb2c6050d13b93660fa4d27d2d28d6eb/pkg/kubelet/util/pod_startup_latency_tracker.go#L181-L189

I think of it as if the pod tries to pull the image even once (so it sets its `firstStartedPulling`), we are in trouble (may get negative values). When it happens, the pod cannot be running unless one of these situations happens:

1. It itself downloads the image and sets its `lastFinishedPulling`.
2. It doesn't succeed, but also it cannot go to a running state before `EnsureImageExists` returns true. When `EnsureImageExists` does that, the pod either successfully downloads the image this time or realizes it exists.

Do you have a subtle corner case in your mind that I'm missing?

bart0sh commented:
@ali-a-a CI failures seem to be related to your changes, at least some of them. PTAL.

ali-a-a commented:
> @ali-a-a CI failures seem to be related to your changes, at least some of them. PTAL.

Yeah, there is a race condition in my unit test. I've gotten a bit busy these days, but I'll fix it shortly.

ali-a-a commented:
I fixed the unit test. Only `pull-kubernetes-unit-windows-master` fails now...

Its log says that one pod doesn't start pulling at all (both `firstStartedPulling` and `lastFinishedPulling` are not set). I conjecture the blocking mechanism in `fake_runtime` doesn't block the pods on windows-master as the first pod pulls the image, although it is blocked. Consequently, the second pod doesn't pull as the image already presents on the machine.

```
 I0423 23:27:45.587568    7940 image_manager.go:130] test-ns/test_pod2/missing_image Pulling image "missing_image"
I0423 23:27:45.587680    7940 image_manager.go:130] test-ns/test_pod1/missing_image Container image "missing_image" already present on machine
    fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 2 times, but was actually called 1 times. 
```

@bart0sh Do you have any idea about it? Moreover, can I simulate the windows-master environment on my local machine? I don't have access to Windows.

bart0sh commented:
@ali-a-a I don't have an idea, but I have a windows machine. I'll try to debug failing test there when I have time, hopefully tomorrow.

bart0sh commented:
/retest

bart0sh commented:
@ali-a-a I could only reproduce this when I created heavy load on my Windows laptop. My guess so far is that CI instances are low-powered and that caused the test to run slower than expected, so go-routines started at line 912 are not able to run fast enough to be registered by fake runtime before the check on line 920.

Calling `wg.Wait()` instead of sleeping for 1 second can help with this.

BTW, the error `fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 2 times, but was actually called 1 times.` seems to be ignored. You should probably consider using assert on line 917 instead of simply calling `fakeRuntime.AssertCallCounts("PullImage", 2)`.

ali-a-a committed:
Fix fake runtime's image pull

Fixed fake_runtime.PullImage to be blocked at the fake image pulling behavior,
not just before before returning

ali-a-a commented:
> @ali-a-a I could only reproduce this when I created heavy load on my Windows laptop. My guess so far is that CI instances are low-powered and that caused the test to run slower than expected, so go-routines started at line 912 are not able to run fast enough to be registered by fake runtime before the check on line 920.

@bart0sh Thank you so much for looking into it. I found the exact problem of CI failures. In `fake_runtime.PullImage`, the blocking mechanism only blocks the function before returning - it doesn't block the actual fake pulling behavior (i.e., `f.ImageList = append(f.ImageList, i)`). It means that in the previous version of the code, one goroutine could pull the image, although it was blocked. I fixed `fake_runtime.PullImage` in my third commit. Now, the whole operation gets blocked not just the return statement.

> Calling `wg.Wait()` instead of sleeping for 1 second can help with this.

It could also cause an issue, but in that case, the other unit test (`TestMaxParallelImagePullsLimit`) would also fail. Moreover, it is not that trivial to use `wg.Wait()` here. As both pull calls are blocked, using wg.Wait() causes deadlock. If tests get flaky because of these test cases, we can think about a solution in another PR probably.

> BTW, the error fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 2 times, but was actually called 1 times. seems to be ignored. You should probably consider using assert on line 917 instead of simply calling fakeRuntime.AssertCallCounts("PullImage", 2).

I double-checked it. When everything is okay, and we change the expected number of calls from 2 to, say, 3, the test blows up. So, it seems that it actually asserts it. Did you mean something else by being ignored?

bart0sh commented:
> Did you mean something else by being ignored?

Sorry for not being clear. Here is what I've seen in the log:
```
    fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 2 times, but was actually called 1 times.
    image_manager_test.go:949: 
        	Error Trace:	C:/kubernetes/pkg/kubelet/images/image_manager_test.go:949
        	Error:      	Should be true
        	Test:       	TestParallelPodPullingTimeRecorderWithErr
```
fake_runtime.go:162 produced the error message, but it was ignored and the test failed at image_manager_test.go:949, which is `assert.True(t, fakePodPullingTimeRecorder.finishedPullingRecorded[pods[0].UID])`. If we add `assert.True(t, fakeRuntime.AssertCallCounts("PullImage", 2))` the error would be caught and the test would fail at image_manager_test.go:946.

ali-a-a commented:
> fake_runtime.go:162 produced the error message, but it was ignored and the test failed at image_manager_test.go:949, which is `assert.True(t, fakePodPullingTimeRecorder.finishedPullingRecorded[pods[0].UID])`. If we add `assert.True(t, fakeRuntime.AssertCallCounts("PullImage", 2))` the error would be caught and the test would fail at image_manager_test.go:946.

Thanks for the clarification. I tested it, and it seems without the `assert.True()`, `fakeRuntime.AssertCallCount` still causes the test to fail. It is what I get when everything is okay and only `fakeRuntime.AssertCallCounts("PullImage", 3)` fails (changed 2 to 3 to make a failure):

```
--- FAIL: TestParallelPodPullingTimeRecorderWithErr (2.00s)
    fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 3 times, but was actually called 2 times.

FAIL    k8s.io/kubernetes/pkg/kubelet/images    8.163s
```

So, the error gets caught there, but it lets the test finish and shows all the failed asserts. Having `assert.True(t, fakeRuntime.AssertCallCounts("PullImage", 3))` only makes it better for tracing. It shows the exact line test failed in the unit test. In terms of causing the test to fail, they act the same.

```
--- FAIL: TestParallelPodPullingTimeRecorderWithErr (2.00s)
    fake_runtime.go:162: AssertCallCounts: expected PullImage to be called 3 times, but was actually called 2 times.
    image_manager_test.go:917: 
                Error Trace:    /home/kos/github/kubernetes/pkg/kubelet/images/image_manager_test.go:917
                Error:          Should be true
                Test:           TestParallelPodPullingTimeRecorderWithErr
```

But if you still think it is better to have `assert.True()`, we can add it.

bart0sh commented on pkg/kubelet/container/testing/fake_runtime.go:
```diff
@@ -318,6 +322,30 @@ func (f *FakeRuntime) GetContainerLogs(_ context.Context, pod *v1.Pod, container
+		// Block the function before adding the image to f.ImageList
+		select {
+		case <-ctx.Done():
+		case <-f.imagePullTokenBucket:
```
Here f is unlocked, but the code uses its field.

ali-a-a replied:
Shouldn't it be like that? It is a channel, and we want all the goroutines to be blocked behind it. So, the channel should not be protected by the lock. The current implementation also uses this field without a lock. I only changed the place where we block the function.

bart0sh replied:
I don't know how it should be. I've just noticed that usage of all other fields is guarded by the lock, but this one is not. So, potentially it can cause a race if this field is accessed by 2 or more go-routines at the same time.

ali-a-a replied:
I see. Thank you for looking at it.

For as long as I can understand, in this part of the code, we want all the go-routines to access the field (i.e., the channel) at the same time so they can sit there and wait for new messages. This is the reason this field is not protected by the lock. If we held the lock while accessing this channel, we could get a deadlock as the channel is exclusively accessed by a go-routine that is waiting for someone else to write to the channel, but it wouldn't happen as no one else could access the channel.

Moreover, even before these changes, this field was accessed without the lock. This PR only changes the location of the select statement.

https://github.com/kubernetes/kubernetes/blob/8aa1d337e955b7fa8d8656b77a1ef56b284f44cf/pkg/kubelet/container/testing/fake_runtime.go#L344-L349

Also, if we had a race condition, we would probably get an error in test results as tests run with the -race flag.

bart0sh commented on pkg/kubelet/container/testing/fake_runtime.go:
```diff
@@ -318,6 +322,30 @@ func (f *FakeRuntime) GetContainerLogs(_ context.Context, pod *v1.Pod, container
+		select {
+		case <-ctx.Done():
+		case <-f.imagePullTokenBucket:
+		case pullImageErr := <-f.imagePullErrBucket:
```
The same here.

ali-a-a replied:
Same as my comment here https://github.com/kubernetes/kubernetes/pull/131324#discussion_r2063840510

bart0sh commented:
/lgtm

/assign @SergeyKanzhelev @mrunalp 
for approval

esotsal commented:
Thanks @bart0sh and @ali-a-a !

ali-a-a commented:
@bart0sh @SergeyKanzhelev @mrunalp 

It seems that this PR has a merge conflict now. If you still plan to review the PR, please let me know, and I'll rebase it.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

great discussions in PR comments

bart0sh commented on pkg/kubelet/util/pod_startup_latency_tracker.go:
```diff
@@ -150,7 +150,9 @@ func (p *basicPodStartupLatencyTracker) RecordImageFinishedPulling(podUID types.
 	}
 
-	state.lastFinishedPulling = p.clock.Now() // Now is always grater than values from the past.
+	if !state.firstStartedPulling.IsZero() {
```
This at least requires an explanation. It looks suspicious that `RecordImageFinishedPulling` silently doesn't do what it's supposed to do.

ali-a-a replied:
It happens because we call `RecordImageFinishedPulling` in `EnsureImageExists` every time the image is already present on the machine. There are two cases in  `EnsureImageExists`. One is that the pod realizes the image exists right away and doesn't start recording. The other one is that it doesn't find the image in its first try, starts recording, goes to the imagePullLoopback phase, finds out that the image exists after a retry, and stops recording. We could distinguish between these cases by adding a new method to the `ImagePodPullingTimeRecorder` interface that tells us whether the recording has started or not. But I think the current implementation keeps it simpler. Right now, we just tell the pod, "Stop recording whenever you find the image". Under the hood, if the recording is going on, we stop it. If it is not, the stop operation is a no-op. This explanation also applies to this comment https://github.com/kubernetes/kubernetes/pull/131324#discussion_r2328688453. But please let me know if you have other ideas.

bart0sh commented:
@ali-a-a Please, squash commits into 2:
- Fix negative pod startup duration
- Fix fake runtime's image pull

Some more info in commit messages would be nice to have as well.
